rfcs/010-gateway-refactor/010-gateway-refactor.md

```diff
@@ -0,0 +1,211 @@
+- Start Date: 09-12-2022
+- Status: Proposed
+
+# Gateway refactor
+
+## Summary
+
+This proposes a full rewrite of the current gateway service. In addition, it also proposes
+changes to the way that access control to Jupyter servers is handled. Currently, each Jupyter
+server establishes a session with the user. But going forward each user will have a session
+with the gateway which will be used to control access into each Jupyter server. 
+The main motivation for changing the access control for Jupyter servers is the desire to run
+each Jupyter server in its own subdomain. The problem with unique subdomains is that each Jupyter
+server's subdomain would have to be registered in Keycloak so that the Oauth2-proxy in each 
+Jupyter server can function properly. But if the Oauth session is handled by the gateway we can
+have a single callback address in Keycloak as we currently do.
+
+## Motivation
+
+Reasons for rewriting the gateway:
+- the current code base is hard to maintain
+- one service could replace both traefik and the gateway
+- problems with managing sessions across different services/clients
+- using golang should result in more efficient use of resources and improved performance
+
+Reasons for centralized session authentication:
+- running sessions in subdomains which would allow for SSH access
+
```

Comment:
Could we add something on Requirements? Does the following seem reasonable?

## Requirements

We currently have two separate issues in which the gateway plays an integral part:
- managing access from external entities (browsers, command line tools) to Renku resources
  - we need to ensure that all requests hitting our internal endpoints have appropriate identities and authorizations
- supporting access from (internal) Renku services to external protected entities
  - in this case we are using tokens generated by external services and managed by Renku to access these resources

In general, we should limit token distribution such that Renku services have them only as necessary, especially for refresh tokens. We should also be aware of the privileges associated with tokens we manage, specifically for tokens relating to external services.

We should understand what happens if we lose all tokens and should be able to recover from this situation. We should also have some basic understanding of the implications of being compromised and what could happen if all tokens which we manage become untrusted.

Reply:
> if we lose all the tokens

@seanrmurphy I am not sure what do you mea by this. Are you referring to the tokens that we expect to be persisted suddenly not being available?

> of the implications of being compromised and what could happen if all tokens which we manage become untrusted.

Can you give an example scenario here? Because this seems really broad in order to address. 

Also there is a difference between the session IDs we will be issuing and the Keycloak or Gitlab access or refresh tokens. By default we do not trust any sessionID - unless we can find it in the database. The keycloak access and refresh tokens can be verified with the keycloak public signing key. The keycloak and the gitlab tokens can also all be introspected with keycloak and gitlab respectively. But if someone gains access to the database where the tokens are stored then there is not much that we can do other than shut everything down and let users know. We will encrypt the stuff in the database. But I think we will have to have the encryption key in the gateway service so that we can actually use the tokens. So it possible that an attack could get access to the encryption key since we will using it all the time anyhow.

Reply:
> @seanrmurphy I am not sure what do you mean by this. Are you referring to the tokens that we expect to be persisted suddenly not being available?

It relates to redis and requirements: we can impose a requirement that tokens be persisted (for how long?) or we can think of disposable tokens. From our discussion earlier, I think that you basically assume tokens will be persisted.

That said, if someone on the operations team screwed up and lost all the tokens, would there be BAD implications? I guess in reality, it would prob mean some friction/inconvenience for users but it would not render the platform unusable.

> Can you give an example scenario here? Because this seems really broad in order to address.

Being compromised is of course a risk and there are many variants of it. To be systematic, we could break it down into different scenarios (which could be worth doing - eg one refresh token is snooped, one access token is snooped, all tokens in one service are snooped etc), but for now, let's take the worst case: an attacker gets access to all the tokens which we are managing - what would be the implications here? I guess she would be able to impersonate all Renku users (on all projects that user has access to?)...would she be able to delete all projects the users have access to (and demand some ransom)? For me, the answer here does not need to be definitive, but we should be aware of and document at least the envelope of the risk profile we consider...

Reply:
So if the tokens are lost and the persisted tokens in the database is lost that would mean that:
- kg integration would fail but for new users who log in it should start to work
- re-provisioning of the kg triples would never complete 100% but for whoever has logged in since it will work
- users would have to log in again in the UI
- the notebook service, the sidecar and core service would also not be able to function fully for existing servers (or cached projects for the core service) but this should in theory be able to work again when users logged back in with the ui

If the attacker gets access to all tokens we manage they would be able to impersonate all users and not just read users' data from a repo but also push. So they could do a series of force pushes to delete data. Perhaps branch protection rules may help here but this is not something we want to be relying on. So the consequences are really bad.

Reply:
> or cached projects for the core service
I don't think this is true. Since we get the token in each request, and a request means a user has logged in again, it shouldn't affect anything. It'd only break e.g. a long running job like dataset import that was started before the issue and tries to push after it.